include/circt/Dialect/FIRRTL/Types.h

```diff
@@ -70,6 +70,10 @@ class FIRRTLType : public Type {
   using Type::Type;
 };
 
+/// Returns whether the two types are equivalent. See the FIRRTL spec for the
+/// full definition of type equivalence.
+bool areTypesEquivalent(FIRRTLType destType, FIRRTLType srcType);
```

Comment:
I would make this comment (and later comments) clear that this is weaker type equivalence than is defined in the FIRRTL spec (namely that this is type equivalence of the passive types).

The spec explicitly calls out that `{a: {flip b: UInt}} is not equal to {flip a: {b: UInt}}`. However, our canonicalization makes these types of statements equivalent (and I'd argue they _are_ equivalent). I'm annoyed that a corollary of this statement is "canonicalization is disallowed".

In discussing this with @mikeurbach (as well as offline with @jackkoenig), I am fine with deviating from the spec here so long as we accept circuits which the Scala FIRRTL Compiler (SFC) rejects, but not if we reject circuits that the SFC accepts. (This assertion holds with this PR.)

Reply:
> In discussing this with @mikeurbach (as well as offline with @jackkoenig), I am fine with deviating from the spec here so long as we accept circuits which the Scala FIRRTL Compiler (SFC) rejects, but not if we reject circuits that the SFC accepts. (This assertion holds with this PR.)

This makes sense to me.

Reply:
Sounds good to me. I updated the comment here and on the specific helper predicate for bundle element equivalence.